src/Tasks/Microsoft.NET.Build.Tasks/JoinItems.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using System.Linq;
+
+namespace Microsoft.NET.Build.Tasks
+{
+    public class JoinItems : TaskBase
```

Comment:
So you're trying to copy metadata from Right to Left items that correspond? You can do that with shared metadata batching:

```xml
<Project>
 <ItemGroup>
  <L Include="foo" />
  <L Include="bar" />
  <L Include="baz" />
  <L Include="bat" />

  <R Include="foo" M="m1" />
  <R Include="bar" M="m2" />
  <R Include="baz" M="m3" />
  <R Include="bat" M="m4" />
 </ItemGroup>

 <Target Name="ExecuteJoin">
  <ItemGroup>
   <Joined Include="@(L)" Condition="'%(Identity)' != '@(R) '">
    <M Condition="'%(L.M)' == ''">@(R->'%(M)')</M>
   </Joined>
  </ItemGroup>

  <Message Text="Joined: @(Joined->'%(Identity), %(M)')"/>
 </Target>
</Project>
```

```
s:\>msbuild S:\msbuild\join.proj
Microsoft (R) Build Engine version 15.3.409.57025 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 8/3/2017 11:20:22 AM.
Project "S:\msbuild\join.proj" on node 1 (default targets).
ExecuteJoin:
  Joined: foo, m1;bar, m2;baz, m3;bat, m4
Done Building Project "S:\msbuild\join.proj" (default targets).
```

Only caveat is you can't use different keys for the left and right item--but you can always first add a JoinKey metadatum to both lists.